src/main/java/at/ac/tuwien/kr/alpha/common/Rule.java

```diff
@@ -0,0 +1,140 @@
+	private final Atom headAtom;
+
+	public Rule(List<Literal> bodyLiterals, Atom headAtom) {
+		this.bodyLiterals = bodyLiterals;
+		this.headAtom = headAtom;
+
+		if (!isSafe()) {
+			// TODO: safety check needs to be adapted to solver what the solver actually understands. Will change in the future, adapt exception message accordingly.
+			throw new RuntimeException("Encountered not safe rule: " + toString()
+				+ "\nNotice: A rule is considered safe if all variables occurring in negative literals, builtin atoms, and the head of the rule also occurr in some positive litera.");
+		}
+	}
+
+	/**
+	 *
+	 * @return a list of all ordinary predicates occurring in the rule (may contain duplicates, does not contain builtin atoms).
+	 */
+	public List<Predicate> getOccurringPredicates() {
+		ArrayList<Predicate> predicateList = new ArrayList<>(bodyLiterals.size() + 1);
+		for (Literal literal : bodyLiterals) {
+			predicateList.add(literal.getPredicate());
+		}
+		if (!isConstraint()) {
+			predicateList.add(headAtom.getPredicate());
+		}
+		return predicateList;
+	}
+
+	/**
+	 * Checks whether a rule is safe.
+	 * The actual safety condition may vary over the next improvements.
+	 * Currently, a rule is safe iff all negated variables and all variables occurring in the
+	 * head also occur in the positive body).
+	 * @return true if this rule is safe.
+	 */
+	private boolean isSafe() {
+		// TODO: do the real check.
+		return true;
+		/*Set<VariableTerm> positiveVariables = new HashSet<>();
+		Set<VariableTerm> builtinVariables = new HashSet<>();
+
+		// Check that all negative variables occur in the positive body.
+		for (Literal literal : bodyLiterals) {
+			// FIXME: The following five lines depend on concrete
+			// implementations of the Atom interface. Not nice.
+			if (literal instanceof BasicAtom) {
+				positiveVariables.addAll(literal.getOccurringVariables());
+			} else if (literal instanceof BuiltinAtom) {
+				builtinVariables.addAll(literal.getOccurringVariables());
+			}
+		}
+
+		for (Atom negAtom : bodyAtomsNegative) {
+			for (VariableTerm term : negAtom.getOccurringVariables()) {
+				if (!positiveVariables.contains(term)) {
+					return false;
+				}
+			}
+		}
+		for (VariableTerm builtinVariable : builtinVariables) {
+			if (!positiveVariables.contains(builtinVariable)) {
+				return false;
+			}
+		}
+
+		// Constraint are safe at this point
+		if (isConstraint()) {
+			return true;
+		}
+
+		// Check that all variables of the head occur in the positive body.
+		List<VariableTerm> headVariables = headAtom.getOccurringVariables();
+		headVariables.removeAll(positiveVariables);
+		return headVariables.isEmpty();
+		*/
+	}
+
+	/**
+	 * Returns the n-th atom in the body of this non-ground rule.
+	 * @param atomPosition 0-based position of the body atom.
+	 * @return
+	 */
+	public Literal getBodyLiteral(int atomPosition) {
+		return bodyLiterals.get(atomPosition);
+	}
+
+	public int getNumBodyAtoms() {
```

Comment:
I'd go for something that uses Java terminology ("size", "count", "length"), e.g. `getBodySize()`.